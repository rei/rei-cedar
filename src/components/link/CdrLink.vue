<script setup>
import { useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses.js';
import propValidator from '../../utils/propValidator.js';

const props = defineProps({
    tag: {
      type: String,
      default: 'a',
    },
    href: {
      type: String,
      default: '#',
    },
    inheritColor: {
      type: Boolean,
      default: false,
    },
    modifier: {
      type: String,
      default: '',
      validator: (value) => propValidator(value, ['', 'standalone']),
    },
    /** @ignore */
    target: String,
    /** @ignore */
    rel: String,
});
const baseClass = 'cdr-link';
const style = useCssModule();

const computedHref = computed(() => (props.tag === 'a' ? props.href : null));
const computedRel = computed(() => {
  if (props.target === '_blank') {
    return props.rel || 'noopener noreferrer';
  }
  return props.rel;
});

const modifierClass = computed(() => props.modifier && `${baseClass}--${props.modifier}`);
const inheritColorClass = computed(() => props.inheritColor && 'cdr-link--inherit-color');
</script>

<template>
  <component
    :is="tag"
    :class="mapClasses(style, baseClass, modifierClass, inheritColorClass)"
    :target="target"
    :rel="computedRel"
    :href="computedHref"
  >
    <slot />
  </component>
</template>

<style lang="scss" module src="./styles/CdrLink.module.scss">
</style>
