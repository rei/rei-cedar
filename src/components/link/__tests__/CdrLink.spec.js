import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import sinon from 'sinon';
import CdrLink from '../CdrLink.vue';

describe('CdrLink', () => {
  describe('basic link component', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrLink);
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('computes the base class correctly', () => {
      expect(wrapper.classes()).toContain('cdr-link');
    });

    describe('when target prop is set', () => {
      beforeEach(() => {
        wrapper.setProps({ target: '_self' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets target attr correctly', () => {
        expect(wrapper.attributes().target).toBe('_self');
      });
    });

    describe('with href prop set', () => {
      beforeEach(() => {
        wrapper.setProps({ href: 'www.rei.com' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets a default href', () => {
        expect(wrapper.attributes().href).toBe('www.rei.com');
      });
    });

    describe('with rel prop set to "nofollow"', () => {
      beforeEach(() => {
        wrapper.setProps({ rel: 'nofollow' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('sets rel attr correctly', () => {
        expect(wrapper.attributes().rel).toBe('nofollow');
      });
    });

    describe('when the target prop has been set to "_blank"', () => {
      beforeEach(() => {
        wrapper.setProps({ target: '_blank' })
      });

      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('computes target="_blank" rel attr correctly', () => {
        expect(wrapper.attributes().rel).toBe('noopener noreferrer');
      });
    });

    describe('with standalone modifier set', () => {
      beforeEach(() => {
        wrapper.setProps({ modifier: 'standalone' })
      });

      it('computes classes correctly for standalone modifier', () => {
        expect(wrapper.classes()).toContain('cdr-link--standalone');
      });
    });

    describe('when the tag has been set to button', () => {
      beforeEach(() => {
        wrapper.setProps({ tag: 'button' })
      });

      it('renders a link with a button element and no href attribute', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('when tag has been set to button', ()=>{
     let wrapper;
     let spy; 

      beforeEach(()=>{
        spy = sinon.spy();
        wrapper = shallowMount(CdrLink, {
          propsData: {
            tag: 'button',
          },
          attrs: {
            onClick: spy
          },
        });
      });

      it('renders a link with a button element and no href attribute', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('emits a click', () => {
        wrapper.trigger('click');
        expect(spy.called).toBeTruthy();
      });
    });
  });
});
