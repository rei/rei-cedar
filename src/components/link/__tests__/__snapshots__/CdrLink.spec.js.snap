// Vitest Snapshot v1

exports[`CdrLink > basic link component > renders correctly 1`] = `
<a
  class="cdr-link"
  href="#"
>
  
  
</a>
`;

exports[`CdrLink > basic link component > when tag has been set to button > renders a link with a button element and no href attribute 1`] = `
<button
  class="cdr-link"
>
  
  
</button>
`;

exports[`CdrLink > basic link component > when target prop is set > renders correctly 1`] = `
<a
  class="cdr-link"
  href="#"
  target="_self"
>
  
  
</a>
`;

exports[`CdrLink > basic link component > when the tag has been set to button > renders a link with a button element and no href attribute 1`] = `
<button
  class="cdr-link"
>
  
  
</button>
`;

exports[`CdrLink > basic link component > when the target prop has been set to "_blank" > renders correctly 1`] = `
<a
  class="cdr-link"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  
  
</a>
`;

exports[`CdrLink > basic link component > with href prop set > renders correctly 1`] = `
<a
  class="cdr-link"
  href="www.rei.com"
>
  
  
</a>
`;

exports[`CdrLink > basic link component > with rel prop set to "nofollow" > renders correctly 1`] = `
<a
  class="cdr-link"
  href="#"
  rel="nofollow"
>
  
  
</a>
`;
